presto-main/src/main/java/com/facebook/presto/operator/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,802 @@
+                rowSizes[i] += fixedWidthBlocksSize;
+            }
+        }
+
+        private int calculateNumberRowsFitInBuffers(List<Integer> variableWidthBlockChannels, int rowSizeForfixedWidthBlocks, int startPosition)
+        {
+            int bytesRemaining = capacity - getSerializedBuffersSizeInBytes();
+
+            if (variableWidthBlockChannels.size() == 0) {
+                // There could be cases where there're 0 blocks in a page but positionCount > 0
+                if (rowSizeForfixedWidthBlocks == 0) {
+                    return positionCount - startPosition;
+                }
+                int positionsFit = max(bytesRemaining / rowSizeForfixedWidthBlocks, 1);
+                bufferFull = (positionsFit <= positionCount - startPosition);
+                return min(positionsFit, positionCount - startPosition);
+            }
+
+            verify(rowSizes != null);
+            for (int i = startPosition; i < positionCount; i++) {
+                bytesRemaining -= rowSizes[i];
+
+                if (bytesRemaining <= 0) {
+                    bufferFull = true;
+                    return max(i - startPosition, 1);
+                }
+            }
+
+            return positionCount - startPosition;
+        }
+
+        private void flush(OutputBuffer outputBuffer)
+        {
+            if (bufferedRowCount == 0) {
+                return;
+            }
+
+            SliceOutput output = new DynamicSliceOutput(getSerializedBuffersSizeInBytes());
+            output.writeInt(blockEncodingBuffers.length);  // channelCount
+
+            for (int i = 0; i < blockEncodingBuffers.length; i++) {
+                blockEncodingBuffers[i].serializeTo(output);
+            }
+
+            SerializedPage serializedPage = serde.serialize(output.slice(), bufferedRowCount);
+            outputBuffer.enqueue(lifespan, partition, Collections.singletonList(serializedPage));
+            pagesAdded.incrementAndGet();
+            rowsAdded.addAndGet(bufferedRowCount);
+            bufferedRowCount = 0;
+        }
+
+        private void ensurePositionsCapacity(int positionCount)
+        {
+            if (positions.length < positionCount) {
+                positions = new int[positionCount];
+            }
+        }
+
+        private void ensureRowSizesCapacityAndInitialize()
+        {
+            if (rowSizes.length <= positionCount) {
+                rowSizes = new int[max(2 * rowSizes.length, positionCount)];
+            }
+            else {
+                Arrays.fill(rowSizes, 0, positionCount, 0);
+            }
+        }
+
+        private int getSerializedBuffersSizeInBytes()
+        {
+            int size = 0;
+            for (int i = 0; i < blockEncodingBuffers.length; i++) {
+                size += blockEncodingBuffers[i].getSerializedSizeInBytes();
+            }
+            return SIZE_OF_INT + size;  // channelCount takes one int
+        }
+    }
+
+    static class DecodedBlockNode
```

Comment:
Given that this class is used by XxxBlockEncodingBuffers, I'd make it a top-level class.